src/uu/touch/src/touch.rs

```diff
@@ -261,7 +262,27 @@ fn parse_timestamp(s: &str) -> FileTime {
     };
 
     match time::strptime(&ts, format) {
-        Ok(tm) => local_tm_to_filetime(to_local(tm)),
+        Ok(tm) => {
+            let mut local = to_local(tm);
+            local.tm_isdst = -1;
+            let ft = local_tm_to_filetime(local);
+
+            // We have to check that ft is valid time. Due to daylight saving
+            // time switch, local time can jump from 1:59 AM to 3:00 AM,
+            // in which case any time between 2:00 AM and 2:59 AM is not valid.
+            // Convert back to local time and see if we got the same value back.
+            let ts = time::Timespec {
+                sec: ft.unix_seconds(),
+                nsec: 0,
+            };
+            let tm2 = time::at(ts);
+            if tm.tm_hour != tm2.tm_hour {
+                show_error!("invalid date format {}", s);
```

Comment:
can you add a test which trigger this error? thanks

Reply:
Initially I did, but then I thought about it so I removed the test. Here's my thinking: this test depends on the local time zone of the machine where the test is executed. For example, where I am (Toronto, EST, which is UTC-4), the switch to DST happened at 2 AM on March 14 2021. So I could make a test where I do touch -m -t 202103140200 to try set the file's modification time to 2 AM and it would file like it does with touch from GNU coreutils:

$ touch -m -t 202103140200 file.txt
touch: invalid date format ‘202103140200’

But this is setting the file's time in the local time of the machine where it runs, so I don't know how to make this test location-independent. For example, if this test is run on any machine in Europe it would probably fail, because European countries didn't switch to DST on March 14.

Reply:
Possibly related to this change: `cargo test` currently fails on my Debian machine with a European time zone (GMT):

```
---- test_du::test_du_time stdout ----
current_directory_resolved: 
run: /<redacted>/src/coreutils/target/debug/coreutils touch -a -t 201505150000 date_test
thread 'test_du::test_du_time' panicked at 'Command was expected to succeed.
stdout = 
 stderr = touch: invalid date format '201505150000'
', tests/common/util.rs:174:9
```

Actually `cargo run touch -a -t xxxxxxxxxx file` fails for *any* valid timestamp, irrespective of it being near daylight saving time changes or not.

Adding debug statements for this code shows :

```
tm = { 
  tm_sec: 0, tm_min: 0, 
  tm_hour: 0, 
  tm_mday: 15, tm_mon: 4, tm_year: 115, tm_wday: 0, tm_yday: 0, tm_isdst: 0, 
  tm_utcoff: 0, 
  tm_nsec: 0 
},

tm2 = Tm {
  tm_sec: 0, tm_min: 0, 
  tm_hour: 1, 
  tm_mday: 15, tm_mon: 4, tm_year: 115, tm_wday: 5, tm_yday: 134, tm_isdst: 0,
  tm_utcoff: 3600, 
  tm_nsec: 0
}
```

Note the difference in `tm_utcoff` between the two, which appears to be the cause of this issue.